lib/bolt/plugin/puppetdb.rb

```diff
@@ -15,9 +16,77 @@ def hooks
         ['lookup_targets']
       end
 
+      def warn_missing_fact(certname, fact)
+        @logger.warn("Could not find fact #{fact} for node #{certname}")
+      end
+
+      def fact_name(raw_fact)
+        fact = raw_fact.split(".")
+        fact.drop(1) if fact[0] == 'facts'
+      end
+
       def lookup_targets(opts)
-        nodes = @puppetdb_client.query_certnames(opts['query'])
-        nodes.map { |certname| { 'uri' => certname } }
+        targets = @puppetdb_client.query_certnames(opts['query'])
+        facts = []
+        facts << fact_name(opts['name']) if opts.key?('name')
+        facts << fact_name(opts['uri']) if opts.key?('uri')
+
+        if opts.key?('config')
+          # This is done in parts instead of in place so that we only need to
+          # make one puppetDB query
+          Bolt::Util.walk_vals(opts['config']) do |value|
+            if value.is_a?(String)
+              facts << fact_name(value)
+            end
+            value
+          end
+        end
+
+        facts.uniq!
+        # Returns {'mycertname' => [{'path' => ['nested', 'fact'], 'value' => val'}], ... }
+        fact_values = facts.empty? ? {} : @puppetdb_client.fact_values(targets, facts)
+
+        targets.map do |certname|
+          target_data = fact_values[certname]
+
+          target = {}
+          if opts.key?('uri')
+            target['uri'] = resolve_facts(opts['uri'], certname, target_data)
+          end
+
+          if opts.key?('name')
+            target['name'] = resolve_facts(opts['name'], certname, target_data)
+          end
+
+          target['uri'] = certname unless opts.key?('name') || opts.key?('uri')
+
+          if opts.key?('config')
+            target['config'] = resolve_facts(opts['config'], certname, target_data)
+          end
+
+          target
+        end.compact
+      end
+
+      def resolve_facts(config, certname, target_data)
+        if target_data.nil?
+          @logger.warn("No facts available for #{certname}")
+          return nil
+        end
+
+        Bolt::Util.walk_vals(config) do |value|
+          if value.is_a?(String)
+            data = target_data.detect { |d| d['path'] == fact_name(value) }
+
+            if data.nil?
+              warn_missing_fact(certname, value)
```

Comment:
This should probably include the fact path.

Reply:
The value here will be what the user input, so `facts.myfact` as opposed to the fact path which is `['myfact']`. I think the former will make more sense to the user - I'm not sure how to rename the var to make that more clear?